Ensure that the task is not in the tasks queue when it is rejected from the executor

ahus1 opened the pull request:
Closes #4900

Motivation:

This is a PR to address #4900. It is rather crude, and uses package visibility. While this does the job to make the assertions, I'd be happy to receive advice on how to change it to match the Vert.x's project way.

Conformance:

You should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md

ahus1 commented:
Updated the PR to ensure proper encapsulation in `TaskQueue` even for tests by adding an `isEmpty()` method.

vietj commented:
can you elaborate when/how this happens ? if a task is rejected it usually means that vertx is closed, unless I am wrong ?

ahus1 commented:
For me, Vert.x is used as part of Quarkus, which is then the base of Keycloak. At the moment I'm running some load tests around Keycloak, and I want to have a first naive load shedding in place.

In a Keycloak setup, Quarkus replaces the executor pool of Vert.x with its own thread pool I can ask Quarkus to limit the backlog of the executor thread pool using a configuration `QUARKUS_THREAD_POOL_QUEUE_SIZE=100`. Once I do this, the thread doesn't allow for an unlimited queue size. If under high load the queue size is exceeded, the requests are rejected. 

I'm discussing more elaborate means of load shedding and not relying on the thread pool queue site. Until that is in place, this is all I have. 

I hope this adds the context the issue is missing.

vietj commented:
thanks for the context

vietj commented on src/main/java/io/vertx/core/impl/TaskQueue.java:
```diff
@@ -125,19 +125,30 @@ public Consumer<Runnable> unschedule() {
     synchronized (tasks) {
-      tasks.add(new ExecuteTask(task, executor));
+      ExecuteTask executeTask = new ExecuteTask(task, executor);
+      tasks.add(executeTask);
```
I think instead since we are in a synchronized block, we could move this addition to the bottom of the block after we are sure the executor accepted the task instead of adding/removing it

ahus1 commented:
@vietj - thank you for the review. I've moved the statement, and also added a comment to state why this is the right place. Please give it another review when you have the time.

vietj commented:
thanks for the contribution, do you mind porting this to the 4.x branch ?

ahus1 commented:
@vietj - thank you for reviewing and merging this PR. Here's the backport for this change to 4.x: #4904
